discovery_engine_core/core/src/engine.rs

```diff
@@ -582,62 +583,104 @@ impl Engine {
+                        url: Url::parse("https://foobar.invalid").unwrap(),
+                        source_domain: "foobar.invalid".into(),
+                        date_published: Utc::now().naive_utc(),
+                        image: None,
+                        rank: 0,
+                        score: None,
+                        country: reacted.market.country_code,
+                        language: reacted.market.lang_code,
+                        topic: "invalid".into()
+                    }
+                };
+            }
+        }
+
+        let market = Market {
+            lang_code: document.resource.language.clone(),
+            country_code: document.resource.country.clone(),
+        };
+
         let mut stacks = self.stacks.write().await;
 
         // update relevance of stack if the reacted document belongs to one
-        if !reacted.stack_id.is_nil() {
-            if let Some(stack) = stacks.get_mut(&reacted.stack_id) {
+        if !document.stack_id.is_nil() {
+            if let Some(stack) = stacks.get_mut(&document.stack_id) {
                 stack.update_relevance(
-                    reacted.reaction,
+                    reaction,
                     self.core_config.max_reactions,
                     self.core_config.incr_reactions,
                 );
-            } else if reacted.stack_id == Exploration::id() {
+            } else if document.stack_id == Exploration::id() {
                 self.exploration_stack.update_relevance(
-                    reacted.reaction,
+                    reaction,
                     self.core_config.max_reactions,
                     self.core_config.incr_reactions,
                 );
             } else {
-                return Err(Error::InvalidStackId(reacted.stack_id));
+                return Err(Error::InvalidStackId(document.stack_id));
             }
         };
 
-        match reacted.reaction {
+        match reaction {
             UserReaction::Positive => {
                 let smbert = &self.smbert;
                 let key_phrases = self
                     .kpe
-                    .run(&reacted.snippet)
+                    .run(&document.resource.snippet)
                     .or_else(|_| {
-                        self.kpe
-                            .run(format!("{} {}", reacted.title, reacted.snippet))
+                        self.kpe.run(format!(
+                            "{} {}",
+                            document.resource.title, document.resource.snippet
+                        ))
                     })
                     .map_or_else(
                         #[allow(clippy::if_not_else)]
                         |_| {
-                            vec![if !reacted.title.is_empty() {
-                                reacted.title.to_string()
+                            vec![if document.resource.title.is_empty() {
+                                document.resource.snippet.clone()
```

Comment:
changed it to a method, but I also don't have a clippy lint :thinking: 

Besides the ones which sneaked on `main` and are fixed in #522  (and another new one not yet fixed in #522 but also not related to this PR).

Reply:
there is no lint, because the non-empty checked item still doesn't come first. if you change it to
```rust
    pub fn title_or_snippet(&self) -> &str {
        if !self.title.is_empty() {
            &self.title
        } else {
            &self.snippet
        }
    }
```
then the lint should appear.

imo both orders are ok, i just made the comment because it was noted before in a different pr.